rosidl_cli/test/rosidl_cli/test_entry_points.py

```diff
@@ -0,0 +1,28 @@
+# Copyright 2021 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pytest
+
+from rosidl_cli.entry_points import normalize_entry_point_specs
+from rosidl_cli.entry_points import SpecifierSet
+
+
+def test_entry_point_spec_normalization():
+    assert normalize_entry_point_specs([]) == {}
+
+    with pytest.raises(ValueError):
+        normalize_entry_point_specs(['name?'])
+
+    specs = normalize_entry_point_specs(['c', 'cpp==0.1.0'])
```

Comment:
What's the motivation for the version specification? IIUC this would mean two different packages could do

```python3
    # First package
    entry_points={
        'rosidl_cli.command.generate.type_extensions': [
            'cpp = first_pkg.v1.CppExtension',
        ],
    # Second package
    entry_points={
        'rosidl_cli.command.generate.type_extensions': [
            'cpp = second_pkg.v2.CppExtension',
        ],
```

but why would this situation arise?

Reply:
> What's the motivation for the version specification?

It's there to:

> - Enforce tooling versioning to ensure API/ABI stability and simplify deprecation cycles

I envision packages pinning plugin versions to guarantee API/ABI stability.

Reply:
> I envision packages pinning plugin versions to guarantee API/ABI stability.

Pin in what context? As in, someone has a package with message files and they want to be sure they're only generated with `cpp==0.1.0`? If they're releasing their package they can pin using the [`version_*` attributes](https://www.ros.org/reps/rep-0140.html#build-depend-multiple) in the `<*depend>` tags of the `package.xml`.

Reply:
> As in, someone has a package with message files and they want to be sure they're only generated with cpp==0.1.0?

Precisely.

> If they're releasing their package they can pin using the version_* attributes in the <*depend> tags of the package.xml.

Hmm, do build tools like `colcon` enforce them? What if there's no `package.xml` e.g. if source code generation is performed outside a ROS package? At any rate, it seems way simpler and cheaper to me to have plugin versions (that ultimately map to that of the package) than to overcome those limitations in ROS' package management system.

Reply:
> Hmm, do build tools like `colcon` enforce them? What if there's no package.xml e.g. if source code generation is performed outside a ROS package?

You're right that a version requirement in the `package.xml` AFAIK is ignored by `colcon`, but I don't see how that or generating messages inside or outside a ROS package would make a user choose to give a version specifier here. If the generator is installed from a deb/rpm and the packaged version doesn't meet a user's requirements, then they need to build the generator from source. If they're building the generator from source, then they get to pick the version by checking out the one they want to build.

Have a look at 4c390031d0c3ca7e80301fa4cd322e701b400ed5 to see how much less code there is when the PEP 440 specifiers are removed.

Reply:
> If the generator is installed from a deb/rpm and the packaged version doesn't meet a user's requirements, then they need to build the generator from source. If they're building the generator from source, then they get to pick the version by checking out the one they want to build.

That's absolutely correct whenever you have full control over sources. But it quickly becomes unmanageable as you integrate third-party code or start building software layers on top.

Let's say you want to use some third-party packages: `foo` and `bar`. Both have the same dependencies at the same version. Both use uses interfaces from a `baz_msgs` package at its `1.0.0` version. And yet, it may not be possible to use them in your workspace, or even together at all. Which `baz_msgs` API/ABI do these packages depend on? Version `1.0.0` only tells half the story. That exact same `baz_msgs` package could exhibit completely different API/ABIs depending on the state of the workspace it was built in.

Today, we "ensure" consistency by resting on the assumption that no API/ABI breaks will (should?) occur within a ROS distro, and thus no API/ABI break will occur in code generated by released generators. In a sense, we're (implicitly) versioning packages by distro: package `foo` depends on package `baz_msgs` at version `1.0.0` in Foxy, but breaks for version `1.0.0` in Rolling because a relevant generator had a major version bump.

IMHO spooky actions at a distance like these are brittle (and break `semver` if I'm not mistaken).

> I don't see how that or generating messages inside or outside a ROS package would make a user choose to give a version specifier here.

By pinning a version specifier, you can make sure interfaces API/ABIs will stay the same (or fail to build with a self explanatory message) regardless of the state of workspace (and underlays).

Reply:
As per offline discussion with @sloretz, the issue I'm trying to address is not exclusive to generators... A C++ package can see its public ABI changed if one of its dependencies changed its own -- all without modifying a single source file. Which kinda explains why our buildfarm has to use package versions, distro names, and build timestamps to ensure builds come out OK. 

It doesn't make sense to have an ad-hoc solution for the `rosidl` pipeline alone. I'll strip away version specifiers. Thanks for the discussion @sloretz !!

Still, we should have a discussion about this @ros2/team. It's not semantic versioning what we do, or at least it's not preventing the dependency hell it is meant to prevent. Relevant discussion at [semver/semver#341](https://github.com/semver/semver/issues/341).

Reply:
See 9939a1bb28cc60ed336d02a21b7e3e47bd303216 for an alternative to https://github.com/ros2/rosidl/commit/4c390031d0c3ca7e80301fa4cd322e701b400ed5 that keeps a few checks (make sure specs are satisfied, complain on duplicate entry points).